### DISCUSSION TITLE
Loading an SVG at a specific size

### DISCUSSION THREAD

**Discussion by tonymarklove** [libvips/libvips#4136] [Q&A] [answered]
Is there a way to load an SVG at a specific size without loss of quality?

### Background

It appears that `svgload` will calculate the size of the image first from any `width` and `height` attributes, and then from the `viewBox`. 

So if I run this code:
(I'm using ruby-vips here, but it should be the same idea everywhere.)

```ruby
image = Vips::Image.svgload(path)
puts "#{image.width}x#{image.height}"
```

on this SVG:

```svg
<svg viewBox="0 0 256 256" width="512" height="512">
  ...
</svg>
```

I get `512x512`. And on this SVG:

```svg
<svg viewBox="0 0 256 256">
  ...
</svg>
```

I get `256x256`

### SVG Scaling

`svgload` has the `dpi` and `scale` options, if I know exactly what scale I need before hand. But let's say that I always want to render at `1024x1024`, but I don't know which of the above images I will receive -- or any other image with a different "size".

There doesn't seem to be a way to call `svgload(path, width: 1024)`. So is there a way that I can set the rendered width or set the `scale` factor after I have "loaded" the SVG but before it is processed?

I'm thinking something like:

```ruby
image = Vips::Image.svgload(path)
image.set_scale(1024 / image.width)
image.pngsave(out_path) # or other processing here
```

Thanks!

**Comment by jcupitt** [comment, marked as answer]
Hello @tonymarklove,

It's very easy (you'll kick yourself). Because libvips is a lazy library, pixels are only ever rendered when they are used. This means you can efficiently load twice, perhaps:

```ruby
target_width = 1024

# get the "natural" size
image = Vips::Image.new_from_file("something.svg")
# load at target size
image = Vips::Image.new_from_file("something.svg", scale: target_width / image.width)
```

The exception is the imagemagick loader which is forced to render the entire image on open, unfortunately. For all other loaders, `_new_from_file()` is always quick, safe, and low-memory. The same trick works for formats like PDF.

You can also just use `thumbnail`, eg.:

```ruby
image = Vips::Image.thumbnail("something.svg", 1024)
```

which does exactly the same thing internally.

**Reply by tonymarklove** [reply]
Hi @jcupitt, thanks for the quick response as always!

Aha! I did consider loading twice as in your first example, but I wasn't sure it would be the best way. Good to know it should be pretty efficient.

The fact that `thumbnail` handles all of this and more (even if it's doing the same thing internally) is great. I think I'll go with that. 👍
